graphql_api/types/test_analytics/test_analytics.py

```diff
@@ -1,24 +1,318 @@
+        raise ValidationError(f"Invalid ordering field: {ordering}")
+
+    if first is not None and last is not None:
+        raise ValidationError("First and last can not be used at the same time")
+
+    if after is not None and before is not None:
+        raise ValidationError("After and before can not be used at the same time")
+
+
+def generate_test_results(
+    ordering: TestResultsOrderingParameter,
+    ordering_direction: OrderingDirection,
+    repoid: int,
+    measurement_interval: MeasurementInterval,
+    *,
+    first: int | None = None,
+    after: str | None = None,
+    last: int | None = None,
+    before: str | None = None,
+    branch: str | None = None,
+    parameter: TestResultsFilterParameter | None = None,
+    testsuites: list[str] | None = None,
+    flags: list[str] | None = None,
+    term: str | None = None,
+) -> TestResultConnection:
+    """
+    Function that retrieves aggregated information about all tests in a given repository, for a given time range, optionally filtered by branch name.
+    The fields it calculates are: the test failure rate, commits where this test failed, last duration and average duration of the test.
+
+    :param repoid: repoid of the repository we want to calculate aggregates for
+    :param branch: optional name of the branch we want to filter on, if this is provided the aggregates calculated will only take into account
+        test instances generated on that branch. By default branches will not be filtered and test instances on all branches wil be taken into
+        account.
+    :param interval: timedelta for filtering test instances used to calculate the aggregates by time, the test instances used will be
+        those with a created at larger than now - interval.
+    :param testsuites: optional list of testsuite names to filter by, this is done via a union
+    :param flags: optional list of flag names to filter by, this is done via a union so if a user specifies multiple flags, we get all tests with any
+        of the flags, not tests that have all of the flags
+    :returns: queryset object containing list of dictionaries of results
+
+    """
+    repo = Repository.objects.get(repoid=repoid)
+    if branch is None:
+        branch = repo.branch
+    interval = measurement_interval.value
+    validate(interval, ordering, ordering_direction, after, before, first, last)
+
+    table = get_results(repoid, branch, interval)
+
+    if table is None:
+        return TestResultConnection(
+            edges=[],
+            total_count=0,
+            page_info={},
+        )
+
+    if term:
+        table = table.filter(pl.col("name").str.starts_with(term))
+
+    if testsuites:
+        table = table.filter(
+            pl.col("testsuite").is_not_null() & pl.col("testsuite").is_in(testsuites)
+        )
+
+    if flags:
+        table = table.filter(
+            pl.col("flags").list.eval(pl.element().is_in(flags)).list.any()
```

Comment:
Claude was telling me we might be able to add these guards to this check, not sure if you think we need em tho

pl.col("flags").is_not_null() &  # handle null rows
pl.col("flags").list.len() > 0 &  # handle empty lists

Reply:
i think the is not null check makes sense